packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +55,101 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string) {
+  const lazyHash = new LazyHash(outdir);
+
+  const templatePathsByPlugin: Map<IValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...templatePaths(construct.outdir));
+      }
+    }
+  });
+
+  const reports: ValidationReport[] = [];
+  for (const [plugin, paths] of templatePathsByPlugin.entries()) {
+    const originalHash = lazyHash.value;
+
+    if (!plugin.isReady()) {
+      const report: ValidationReport = {
+        pluginName: plugin.name,
+        success: false,
+        violations: [],
+        metadata: {
+          error: `Validation plugin '${plugin.name}' is not ready`,
+        },
+      };
+      reports.push(report);
+    } else {
+      const report = plugin.validate({ templatePaths: paths });
+      reports.push(report);
+    }
+    if (computeChecksumOfFolder(outdir) !== originalHash) {
```

Comment:
So this prevents the cfn-guard plugin for example from adding suppression data into the template, which is a use-case discussed [here](https://github.com/aws/aws-cdk-rfcs/pull/478#discussion_r1118694621). Any ways around it? or am I missing something?

Reply:
That use-case in the RFC would have had the user adding suppression data via a mechanism that worked prior to synth. It would be the CDK adding the metadata to the template, not the plugin.